File: CONCEPTS/reshaping.py
import pandas as pd
import sqlite3
import os

# Connect to database with error handling
db_path = os.path.join(os.getenv("EXAMPLES", ""), "Boxscores.db")
conn = sqlite3.connect(db_path)

# 1. Bring into Boxscores the player and team names via SQL (mimics the lazy join)
# Using SQL for the initial join is more memory-efficient, similar to dbplyr/tbl
query_sql = """
SELECT b.*, p.PLAYER_NAME, t.TEAM_NAME
FROM Boxscores b
LEFT JOIN Players p ON b.PLAYER_ID = p.PLAYER_ID
LEFT JOIN Teams t ON b.TEAM_ID = t.TEAM_ID
ORDER BY b.GAME_ID, b.TEAM_ID
"""

shooting = pd.read_sql_query(query_sql, conn)

# Drop ID columns to match the R select(-PLAYER_ID, -TEAM_ID)
shooting = shooting.drop(columns=['PLAYER_ID', 'TEAM_ID'])

print(shooting)
conn.close()

# 2. Pivot longer to stack all shot types
# Equivalent to pivot_longer
shooting_long = shooting.melt(
    id_vars=[col for col in shooting.columns 
        if col not in ['FGM', 'FGA', 'FG3M', 'FG3A', 'FTM', 'FTA']],
    value_vars=['FGM', 'FGA', 'FG3M', 'FG3A', 'FTM', 'FTA'],
    var_name='stat_type',
    value_name='count'
)

# 3. Calculate shooting percentages by player
# Extract 'shot_type' (FG, FG3, FT) and 'made_attempt' (M, A)
shooting_long['shot_type'] = shooting_long['stat_type'].str[:-1]
shooting_long['made_attempt'] = shooting_long['stat_type'].str[-1]

# Pivot wider to get 'M' and 'A' as columns
shooting_summary = shooting_long.pivot_table(
    index=['PLAYER_NAME', 'shot_type'],
    columns='made_attempt',
    values='count',
    aggfunc='sum'
).reset_index()

# Calculate percentages and summarize
shooting_summary['pct'] = (shooting_summary['M'] / 
                           shooting_summary['A']) * 100

# group_by and summarise
final_summary = shooting_summary.groupby(['PLAYER_NAME', 'shot_type']).agg(
    total_attempts=('A', 'sum'),
    shooting_pct=('pct', 'mean')
).reset_index()

print(final_summary)

File: CONCEPTS/reshaping.R
library(tidyverse)
library(RSQLite)

# Connect to database with error handling
path_to_database <- file.path(Sys.getenv("EXAMPLES"), "Boxscores.db")
 
con <- dbConnect(RSQLite::SQLite(), path_to_database)

# Bring into Boxscores the player and team names
query <- tbl(con, "Boxscores")                     |>
  left_join(tbl(con, "Players"), by = "PLAYER_ID") |>
    left_join(tbl(con, "Teams"), by = "TEAM_ID")   |>
        arrange(GAME_ID, TEAM_ID)                  |>
          select(-PLAYER_ID, -TEAM_ID)             |>
            as_tibble() 
           

# Run the query
shooting <- query |>
  collect()

print(shooting)

dbDisconnect(con)

# Pivot longer to stack all shot types
shooting_long <- shooting |>
  pivot_longer(
    cols = c(FGM, FGA, FG3M, FG3A, FTM, FTA),
    names_to = "stat_type",
    values_to = "count"
  )

# Calculate shooting percentages by player
shooting_summary <- shooting_long |>
  separate(stat_type, into = c("shot_type", "made_attempt"), sep = -1) |>
  pivot_wider(
    names_from = made_attempt,
    values_from = count,
    values_fn = sum
  ) |>
  mutate(pct = M / A * 100) |>
  group_by(PLAYER_NAME, shot_type) |>
  summarise(
    total_attempts = sum(A),
    shooting_pct = mean(pct, na.rm = TRUE),
    .groups = "drop"
  )

print(shooting_summary)